### PR TITLE
Add an event to override soldiers' class for their first promotion

### LINKED ISSUE #801: Add an Event to arbitrarily set soldier class when they are first promoted
Adding an Event to the start of `XCGS_Unit::RankUpSoldier` would allow mods setting soldier's class based on arbitrary conditions when they are first promoted.

For example, when used alongside Not Created Equal, it would be possible to make high aim soldiers gravitate towards becoming snipers.

### PR DESCRIPTION
**Pull request by Iridar**
Fixes #801

### CODE REVIEW

**Comment by Xymanek** [comment]
Could you please extract the event trigger to a separate private function?

**Comment by Iridar** [comment]
> Could you please extract the event trigger to a separate private function?

Sure, why?

**Comment by Xymanek** [comment]
It's cleaner, doesn't pollute the original code with the tuple setup and easier to see what was changed

**Comment by Iridar** [comment]
Never seen this done for any other highlander event, is this new direction?

**Comment by Xymanek** [comment]
It's done like that for a while now, afaik since 4xx. For examples see https://github.com/X2CommunityCore/X2WOTCCommunityHighlander/pull/793/files or https://github.com/X2CommunityCore/X2WOTCCommunityHighlander/pull/780/files

**Comment by Iridar** [comment]
This PR is ready for merging, but I'm no longer sure it's necessary, because as I realized it's just a more complicated Commander's Choice. Closing it for now.

**Comment by Iridar** [comment]
Reopening because Commander's Choice does big bad voodoo.

**Comment by Iridar** [comment]
> The name of the event is too ambiguous. Something like `FirstPromotionOverrideClass` would represent the purpose much better

I called it PreUnitRankUp because it triggers before UnitRankUp event. 
Renamed it per your request.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 1]
```diff
@@ -12544,22 +12544,22 @@ function RankUpSoldier(XComGameState NewGameState, optional name SoldierClass, o
 }
 
 //	Begin issue #801
-/// HL-Docs: feature:PreUnitRankUp; issue:801; tags:strategy
-/// The `XComGameState_Unit::RankUpSoldier` triggers a `'PreUnitRankUp'` event, allowing mods to override the soldier class template name
+/// HL-Docs: feature:FirstPromotionOverrideClass; issue:801; tags:strategy
+/// The `XComGameState_Unit::RankUpSoldier` triggers a `'FirstPromotionOverrideClass'` event, allowing mods to override the soldier class template name
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc]: Did you mean to put `''` here? Looks weird:
image

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 2]
```diff
@@ -12539,6 +12543,65 @@ function RankUpSoldier(XComGameState NewGameState, optional name SoldierClass, o
...
+///	Example of an Event Listener Function:
+///	```unrealscript
+///	static function EventListenerReturn ListenerEventFunction(Object EventData, Object EventSource, XComGameState NewGameState, Name Event, Object CallbackData)
+///	{
+///		local XComLWTuple			Tuple;
+///		local XComGameState_Unit	UnitState;
+///
+///		Tuple = XComLWTuple(EventData);
+///		UnitState = XComGameState_Unit(EventSource);
+///		if (Tule == none || UnitState == none)
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc]:
```suggestion
///		if (Tuple == none || UnitState == none)
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc (hunk 1 above)
Comment (Iridar) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc]: To mark it as a `'Name'`.

Comment (Xymanek) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc]: I don't think it's necessary - you already have the backticks highlighting and everyone knows that event names are always of type `name`

Comment (Iridar) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc]: I'm gonna comply with all your requests, but just for the record, I think this level of nitpicking is ridiculous.

**Comment by Xymanek** [comment]
This should probably be given a more appropriate title and squash merged (when time comes)

**Comment by Iridar** [comment]
> This should probably be given a more appropriate title and squash merged (when time comes)

Renamed. I dunno how to squish, though.

**Comment by Iridar** [comment]
Approved by @Musashi1584 , so I assume it's ready to be merged.

**Pull request merged**